apollo-ast/src/main/kotlin/com/apollographql/apollo3/ast/internal/SchemaValidationScope.kt

```diff
@@ -149,6 +203,167 @@ internal fun ValidationScope.syntheticSchemaDefinition(): GQLSchemaDefinition {
   )
 }
 
+/**
+ * @param newNames a mapping from the foreigne name to the new name
+ * @param directivesToStrip the directives to strip (using new names)
+ */
+private class ForeignSchema(
+    val name: String,
+    val definitions: List<GQLDefinition>,
+    val newNames: Map<String, String>,
+    val directivesToStrip: List<String>
+)
+
+/**
+ * Parses the schema extensions
+ *
+ * Example: extend schema @link(url: "https://specs.apollo.dev/kotlin_labs/v0.1", import: ["@nonnull"])
+ */
+private fun List<GQLSchemaExtension>.getForeignSchemas(
+    issues: MutableList<Issue>,
+    builtinDefinitions: List<GQLDefinition>,
+): List<ForeignSchema> {
+  val schemaExtensions = this
+
+  val foreignSchemas = mutableListOf<ForeignSchema>()
+
+  schemaExtensions.forEach { schemaExtension ->
+    schemaExtension.directives.forEach eachDirective@{ gqlDirective ->
+      if (gqlDirective.name == "link") {
+        val minimalSchema = builtinDefinitions + linkDefinitions()
+        val scope = DefaultValidationScope(
+            minimalSchema.filterIsInstance<GQLTypeDefinition>().associateBy { it.name },
+            minimalSchema.filterIsInstance<GQLDirectiveDefinition>().associateBy { it.name },
+        )
+        scope.validateDirective(gqlDirective, schemaExtension) {
+          issues.add(it.constContextError())
+        }
+        if (scope.issues.isNotEmpty()) {
+          issues.addAll(scope.issues)
+          return@eachDirective
+        }
+
+        val arguments = gqlDirective.arguments!!.arguments
+        val url = (arguments.first { it.name == "url" }.value as GQLStringValue).value
+        var prefix = (arguments.firstOrNull { it.name == "as" }?.value as GQLStringValue?)?.value
+        val import = (arguments.firstOrNull { it.name == "import" }?.value as GQLListValue?)?.values
+
+        var components = url.split("/")
+        if (components.last().isBlank()) {
+          // https://spec.example.com/a/b/c/mySchemanameidentity/v1.0
+          components = components.dropLast(1)
+        } else if (components.last().startsWith("?")) {
+          // https://spec.example.com/a/b/c/mySchemanameidentity/v1.0/?key=val&k2=v2#frag
+          components = components.dropLast(1)
+        }
+
+        if (components.size < 2) {
+          issues.add(Issue.ValidationError("Invalid @link url: 'url'", gqlDirective.sourceLocation))
+          return@eachDirective
+        }
+
+        val foreignName = components[components.size - 2]
+
+        if (prefix == null) {
+          prefix = foreignName
```

Comment:
If I understand correctly it means that `extend schema @link(url: "https://specs.apollo.dev/kotlin_labs/v0.1", import: ["@nonnull"])` won't import `@nonnull` but `@kotlin_labs__nonnull`?

Reply:
Not exactly:

```graphql
# import removes the namespace
extend schema @link(url: "https://specs.apollo.dev/kotlin_labs/v0.1", import: ["@nonnull"])

type Foo {
  field: String! @nonnull
}
```

```graphql
# without import uses the default namespace
extend schema @link(url: "https://specs.apollo.dev/kotlin_labs/v0.1")

type Foo {
  field: String! @kotlin_labs__nonnull
}
```

```graphql
# as changes the namespace
extend schema @link(url: "https://specs.apollo.dev/kotlin_labs/v0.1", as: "apollo")

type Foo {
  field: String! @apollo__nonnull
}
```

Reply:
Ah ok got it! 👍  Makes sense actually 😊

Reply:
I think the confusing part is that "import" is more about "rename" than "import" but once you get past that, it all makes sense